### DISCUSSION TITLE
client-side newtype validation

### DISCUSSION THREAD

**Discussion by clux** [kube-rs/kube#654] [Ideas]
A problem that came up in #653 in https://github.com/kube-rs/kube-rs/pull/653#discussion_r730389163 is about **if** /  **where** / and **how** we should do client side validation.

In that PR we went with:
- yes, we should do it when wrapping kubernetes apis (api params that we hand roll + runtime abstractions)
- we should do it through own error type

Client-side validation is not something we can truly keep up to date with kubernetes code since it's not found in the protos or the swagger.json.

The [amount of code for each newtype](https://github.com/kube-rs/kube-rs/blob/14033ba1ac723d1be88ed921c69b00b39495f6d9/kube-runtime/src/events/event.rs#L99-L161) is a bit troubling to me though. Particularly given the uncertainty of longevity.

So question is, should we scrap this. Or should we improve this? Some things I can think to improve it would be:

- derive-more for some of the impls
- snafu for the error type
- macroify the pattern

Ultimately there's only 3 bits of information encoded here: error name, error string, validation.

**Comment by nightkr** [comment]
I'd say no, validation is generally the server's responsibility (unless it
causes, say, a serialization problem locally).
